react/src/lib/components/DeckGLMap/layers/axes2d/axes2DLayer.ts

```diff
@@ -0,0 +1,448 @@
+import {
+    COORDINATE_SYSTEM,
+    Color,
+    CompositeLayer,
+    Viewport,
+    UpdateParameters,
+    LayersList,
+    OrthographicViewport,
+} from "@deck.gl/core/typed";
+import BoxLayer from "./boxLayer";
+import { Position3D, ExtendedLayerProps } from "../utils/layerTools";
+import { layersDefaultProps } from "../layersDefaultProps";
+import { TextLayer } from "@deck.gl/layers/typed";
+
+export interface Axes2DLayerProps<D> extends ExtendedLayerProps<D> {
+    labelColor?: Color;
+    labelFontSize?: number;
+    fontFamily?: string;
+    axisColor?: Color;
+    marginH: number;
+    marginV: number;
+}
+
+type TextLayerData = {
+    label: string;
+    from: Position3D; // tick line start
+    to: Position3D; // tick line end
+    size: number; // font size
+};
+
+export default class Axes2DLayer extends CompositeLayer<
+    Axes2DLayerProps<unknown>
+> {
+    shouldUpdateState({
+        props,
+        oldProps,
+        context,
+        changeFlags,
+    }: UpdateParameters<this>): boolean {
+        return (
+            super.shouldUpdateState({
+                props,
+                oldProps,
+                context,
+                changeFlags,
+            }) || changeFlags.viewportChanged
+        );
+    }
+
+    getAnchor(d: TextLayerData): string {
+        const is_xaxis = d.from[1] !== d.to[1];
+        if (is_xaxis) {
+            return "middle";
+        }
+
+        const screen_from = this.context.viewport.project(d.from);
+        const screen_to = this.context.viewport.project(d.to);
+        const is_labels = d.label !== "X" && d.label !== "Y" && d.label !== "Z"; // labels on axis or XYZ annotations
+        if (is_labels) {
+            if (screen_from[0] < screen_to[0]) {
+                return "start";
+            }
+        }
+
+        return "end";
+    }
+
+    getLabelPosition(d: TextLayerData): Position3D {
+        const is_labels = d.label !== "X" && d.label !== "Y" && d.label !== "Z"; // labels on axis or XYZ annotations
+        if (is_labels) {
+            const tick_vec = [d.to[0] - d.from[0], d.to[1] - d.from[1]];
+            if (d.to[2] && d.from[2]) tick_vec.push(d.to[2] - d.from[2]);
+
+            const s = 0.5;
+            return [
+                d.to[0] + s * tick_vec[0],
+                d.to[1] + s * tick_vec[1],
+                d.to[2] + s * tick_vec[2],
+            ];
+        } else {
+            // XYZ axis annotaion.
+            return d.to;
+        }
+    }
+
+    getBaseLine(d: TextLayerData): string {
+        const is_x_annotaion = d.label === "X";
+        if (is_x_annotaion) {
+            return "center";
+        }
+
+        const is_xaxis_label = d.from[1] !== d.to[1];
+        return is_xaxis_label ? "top" : "center";
+    }
+
+    renderLayers(): LayersList {
+        const vpBounds = this.context.viewport.getBounds();
+
+        const mh = this.props.marginH / 100; // to percentage.
```

Comment:
Margin specified as percentage doesn't seem to work well with resizing. Would it be better to specify in pixels?